### DISCUSSION TITLE
Need help to make a custom kitten (I'm a noob)

### DISCUSSION THREAD

**Discussion by j0k4rX** [kovidgoyal/kitty#4363] [Q&A] [answered]
Recently I've made an issue
[When there are no tabs, with the same command.. Go to next window if exists ](https://github.com/kovidgoyal/kitty/issues/4361#issue-1081103133)

kovidgoyal suggested me to make a custom kitten for the above feature.

I'm a beginner and don't know any programming language.
Therefore, I'll be grateful if you folks help me implement this :)

**Comment by page-down** [comment, marked as answer]
If you don't mind learning some programming, I can guide you on how to write it.

https://docs.python.org/3/tutorial/

You need to use `no_ui` mode for kitten, you can simply follow (copy) the example `zoom_toggle.py`. Create the file `next_tab_or_window.py` in the `~/.config/kitty` folder.

https://sw.kovidgoyal.net/kitty/kittens/custom/#using-kittens-to-script-kitty-without-any-terminal-ui

You need `boss.active_tab_manager` to check the number of tabs.
If the number of `tabs` lists in the Tab Manager is greater than 1, the `next_tab()` method can be executed to switch to the next Tab.
Otherwise, you can use `boss.active_tab` (in the `zoom_toggle.py` example) to call `tab.next_window()` to switch to the next kitty window.

You can run the command `kitty -o 'map ctrl+tab kitten next_tab_or_window.py'` to test in the newly opened kitty instance if it functions properly and if there are any errors that need to be fixed.

The relevant classes and methods are in the `kitty/boss.py` and `kitty/tabs.py` source files.

Without strict variable checking, the effective lines of code required for this function is one line.

**Reply by j0k4rX** [reply]
@page-down I came up with this but still doesn't work though :(
Kindly fix this please
```
def main(args):
    pass

from kittens.tui.handler import result_handler
@result_handler(no_ui=True)
def handle_result(args, answer, target_window_id, boss):
    tab = boss.active_tab
    if tab is not None:
        if boss.active_tab == 0:
            next_window()
        else:
            next_tab()
```

**Reply by j0k4rX** [reply]
Even tried this
```
def handle_result(args, answer, target_window_id, boss):
    tab = boss.active_tab
    if boss.active_tab > 1:
            next_tab()
    if boss.active_tab == 1:
            next_window()
```
But i'm getting this when i press `ctrl+tab`
kitty is doing normal `tab` command

![Screenshot at 2021-12-16 19-55-27](https://user-images.githubusercontent.com/71499522/146385126-2a23a949-cba9-42a7-a502-f95f6eda4a2b.png)

**Reply by page-down** [reply]
First you need the tab manager, a class for managing tabs.
```python
tab_manager = boss.active_tab_manager
if tab_manager is not None:
    # ...
```

There is a `tabs` attribute in the tab manager class. You can use Python's `len()` to get the number of elements in a list, so you can know the number of tabs. 
https://docs.python.org/3/library/functions.html#len

You want to switch to the next tab when there are multiple tabs. `if len(tab_manager.tabs) > 1:`

The class instance should be used to call its methods. That's `tab_manager.next_tab()` and `tab.next_window()`.
https://docs.python.org/3/tutorial/classes.html#class-and-instance-variables

Just take a look at the `tabs.py` source code and do a quick search for `class Tab:`, `class TabManager:`, they are all in there.
https://github.com/kovidgoyal/kitty/blob/master/kitty/tabs.py

In this case you can use `if`, `else` and that's it.

```python
if len(tab_manager.tabs) > 1:
    tab_manager.next_tab()
else:
    tab.next_window()
```

**Reply by j0k4rX** [reply]
i came upto this. yet it doesn't work
what am i missing?

```
def main(args):
    pass

from kittens.tui.handler import result_handler
@result_handler(no_ui=True)
def handle_result(args, answer, target_window_id, boss):
    tab_manager = boss.active_tab_manager
    if tab_manager is not None:
        if len(tab_manager.tabs) > 1:
            tab_manager.next_tab()
        else:
            tab.next_window()
```

**Reply by page-down** [reply]
Did you miss your `tab`?

```python
else:
    tab = boss.active_tab
    if tab is not None:
```

**Reply by j0k4rX** [reply]
ah.. plz have a look at my code
I'm confused. 
Could you plz re-write the program? 

```
def main(args):
    pass

from kittens.tui.handler import result_handler
@result_handler(no_ui=True)
def handle_result(args, answer, target_window_id, boss):
    tab_manager = boss.active_tab_manager
    if tab_manager is not None:
        if len(tab_manager.tabs) > 1:
            tab_manager.next_tab()
        else:
            tab.next_window()
```
```
    else:
    tab = boss.active_tab
    if tab is not None:
```

**Reply by page-down** [reply]
Python programs need to be properly indented. `tab` variable (instance of the Tab class) need to be assigned before calling.
Don't you want to switch to the next tab when the number of tabs is greater than 1, otherwise switch to the next window? Then write this logic in the code.

By now, I have provided all the information needed. Please take your time. You will get there.

**Reply by j0k4rX** [reply]
i couldn't make it work :cry:

here's the code & logic
```
def main(args):
    pass

from kittens.tui.handler import result_handler
@result_handler(no_ui=True)
def handle_result(args, answer, target_window_id, boss):
    tab = boss.active_tab
    tab_manager = boss.active_tab_manager
    if tab_manager is not None:
        if len(tab_manager.tabs) > 1:
            tab_manager.next_tab()
        else:
            tab.next_window()
```

**Reply by page-down** [reply]
I tried your code and this seems to work fine.

```shell
kitty --config=NONE -o 'map ctrl+tab kitten next_tab_or_window.py'
```

You can switch to the next Tab or kitty window normally after pressing `ctrl+tab`.

If you are not talking about a kitty window (panel), but an OS window, you need the function `focus_os_window`. You can refer to the following code.

https://github.com/kovidgoyal/kitty/pull/4316/files

I can't write your code for you. You need to learn some Python. Or wait for someone else.

**Reply by j0k4rX** [reply]
it's true there is no bug in the code
but it doesn't switch to next_tab .. let alone next_window

PS: I know a little c++. I have logic and everything but i can't seem to make the python code compatible with kitty 
I'm on parrot os with kitty v0.19.3

**Reply by page-down** [reply]
This version seems to be a bit old. Please go to the official website to install the latest version. You can also install it manually (extract tarball). You can also download the nightly version from GitHub Releases to try it out.

https://sw.kovidgoyal.net/kitty/binary/

Use the latest version and you should be fine.

**Reply by j0k4rX** [reply]
> i couldn't make it work 😢
> 
> here's the code & logic
> 
> ```
> def main(args):
>     pass
> 
> from kittens.tui.handler import result_handler
> @result_handler(no_ui=True)
> def handle_result(args, answer, target_window_id, boss):
>     tab = boss.active_tab
>     tab_manager = boss.active_tab_manager
>     if tab_manager is not None:
>         if len(tab_manager.tabs) > 1:
>             tab_manager.next_tab()
>         else:
>             tab.next_window()
> ```

does this work in your machine?

**Reply by page-down** [reply]
@j0k4rX 
> does this work in your machine?

I already replied to you 17 days ago "I tried your code and this seems to work fine." and I have nothing more to say. Please read my reply.

You need to express whether you tried to use the latest kitty version, listing each step you took.
Whether or not it still doesn't work after making the attempt.
If you don't specify, I don't know how to help you.

Because the code you have written works fine with me and the logic is basically correct.

**Reply by j0k4rX** [reply]
@page-down Yes, i have tried with the latest version
and it doesn't work

**Reply by page-down** [reply]
Using kitty version 0.24.0, copy your code to `~/.config/kitty/kitten_switch_tab_or_win.py` and run the following command to start kitty for testing.

```shell
kitty --config=NONE -o 'map ctrl+tab kitten kitten_switch_tab_or_win.py'
```

- Press ctrl+shift+enter to create a new kitty window, and press ctrl+tab to switch between windows.
- Press ctrl+shift+t to create a new tab, ctrl+tab will switch between different tabs.

Everything works fine.

As for whether there are other problems on your environment, only you can troubleshoot the problems on your computer. 

At this point, I've tried to be nice and have done everything I can to help you. There is nothing else I can do.

**Reply by j0k4rX** [reply]
Ahh.. thank you
seems it was a computer problem after all
again sorry for disturbing you

**Reply by j0k4rX** [reply]
> i couldn't make it work cry
> 
> here's the code & logic
> 
> ```
> def main(args):
>     pass
> 
> from kittens.tui.handler import result_handler
> @result_handler(no_ui=True)
> def handle_result(args, answer, target_window_id, boss):
>     tab = boss.active_tab
>     tab_manager = boss.active_tab_manager
>     if tab_manager is not None:
>         if len(tab_manager.tabs) > 1:
>             tab_manager.next_tab()
>         else:
>             tab.next_window()
> ```



> I tried your code and this seems to work fine.
> 
> ```shell
> kitty --config=NONE -o 'map ctrl+tab kitten next_tab_or_window.py'
> ```
> 
> You can switch to the next Tab or kitty window normally after pressing `ctrl+tab`.
> 
> If you are not talking about a kitty window (panel), but an OS window, you need the function `focus_os_window`. You can refer to the following code.
> 
> https://github.com/kovidgoyal/kitty/pull/4316/files
> 
> I can't write your code for you. You need to learn some Python. Or wait for someone else.

@page-down it doesn't work in latest version 0.26.3
please help

**Reply by page-down** [reply]
> it doesn't work in latest version 0.26.3

It works fine in 0.26.3. And I don't believe there are any relevant code changes in the new version.

> please help

"please help" does not bring any additional information and I can't make any more guesses or guidance.
Check your environment and start from scratch. E.g. `kitty --config=NONE ...`
See the detailed steps in my reply above on January 4.

**Reply by j0k4rX** [reply]
I needed it to be 
- go to next window
- if windows don't exist, go to next tab

but it works the other way round

FYI, the no ui template has been changed
https://sw.kovidgoyal.net/kitty/kittens/custom/#using-kittens-to-script-kitty-without-any-terminal-ui

**Reply by page-down** [reply]
> I needed it to be ... but it works the other way round

This does not match the description of this topic, which you posted last year, and your code was written based on previous needs.

> Dec 16, 2021
> Recently I've made an issue ...
> ...
> but in case of tabs and windows, the command will switch to next tab
> when no tabs available, the hotkey will trigger go to next window

So change your code logic, check whether there are multiple windows first (`len(tab.windows) > 1`).
Changing an `if` and swapping the actions, very straightforward.

> FYI, the no ui template has been changed ...

The documentation recommends using Python type hints, and there are no backward compatibility issues.
https://docs.python.org/3/library/typing.html

**Reply by j0k4rX** [reply]
Yeah, i mistook tabs for windows. Silly of me!
Thanks.
Finally, i made it work.
